django/db/models/query_utils.py

```diff
@@ -68,6 +68,11 @@ def __invert__(self):
         obj.negate()
         return obj
 
+    def __copy__(self):
+        obj = self._new_instance()
```

Comment:
In this case you may as well do `obj = self.__class__()` as you're not passing anything in.

`Node._new_instance()` only exists to help smooth out the incompatibilities between `Q.__init__()` and `Node.__init__()` when needing to reconstruct objects internally. In fact, it would be nice to eliminate it if possible.

Reply:
You'd think so, but that wasn't the finding I saw, I expect because of the super() chaining, sorting and double-copying an empty list, `self.__class__()` is basically the same a doing `type(x)()` and is just as slow.

Back of the napkin fresh timings, rather than earlier ones (which I didn't save):
```
In [1]: from django.db.models.query_utils import Q
In [2]: x = Q()
In [3]: %timeit type(x)()
1.11 µs ± 11.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
In [4]: %timeit x.__class__()
1.05 µs ± 5.39 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
In [5]: %timeit x._new_instance()
627 ns ± 2.19 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
In [6]: %%timeit
   ...: obj = Node()
   ...: obj.__class__ = Q
430 ns ± 3.96 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
The fastest way (as alluded to in the ticket, I think?) is to inline the class swap and avoid the method call entirely, but at least using `_new_instance` is consistent with both Node's and WhereNode's internal usage.

I'd be happy to see the `__init__` for Q changed, I do have an idea for changes I want to propose for that, but it's a separate ticket and may take some convincing ;)

Reply:
I guess this point became moot when I moved to `Node.copy()` and merged `Node._new_instance()` into it.

Reply:
Guhhh, GitHub notifications took me straight to the line comments, so I didn't even see you had put together an alternate branch. Perhaps I should check it out huh? :)

Reply:
I'm leaning toward reintroducing `Node._new_instance()`, although probably call it `.create()` and make sure it is used everywhere which it hasn't been up to new. The reason for this is that sometimes using `.copy()` has unnecessary overhead of copying `__dict__` when we don't need to.

Reply:
Ah so in your branch's variation, `__copy__` is fast because you inlined the class reassignment trick, which is the fastest possible version I found (being 100-150ns faster than using `_new_instance`), so that's cool.

But `_combine` is no faster than the baseline because of using `self.__class__()` directly (which is the aforementioned `super() chaining, sorting and double-copying`)

Back of the napkin time:

Baseline (main):
```
In [2]: full = Q(a=1, b=2, c=3)
In [3]: full2 = Q(d=4, e=5, f=6)
In [4]: %timeit full & full2
2.76 µs ± 45 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

This branch:
```
In [2]: full = Q(a=1, b=2, c=3)
In [3]: full2 = Q(d=4, e=5, f=6)
In [4]: %timeit full & full2
2.24 µs ± 57.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Your revised branch:
```
In [2]: full = Q(a=1, b=2, c=3)
In [3]: full2 = Q(d=4, e=5, f=6)
In [4]: %timeit full & full2
2.9 µs ± 68.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

I'll happily assume that 2.76us == 2.9us and it's just standard laptop behaviour causing the discrepancy between baseline ~ yours, but I think the class reassignment is still accounting for the difference in this branch.

Reply:
Just re-ran the above tests with your branch again, with the `create` method back in place doing the trick, and that brings combine back down to `2.26µs` (well within the margin of error), FWIW.